src/libm/df.h

```diff
@@ -464,3 +464,13 @@ static INLINE CONST vfloat2 dfsqrt_vf2_vf(vfloat d) {
   vfloat t = vsqrt_vf_vf(d);
   return dfscale_vf2_vf2_vf(dfmul_vf2_vf2_vf2(dfadd2_vf2_vf_vf2(d, dfmul_vf2_vf_vf(t, t)), dfrec_vf2_vf(t)), vcast_vf_f(0.5f));
 }
+
+#ifndef SLEEF_SINGLE_MINMAXNUM_AVAILABLE
+static INLINE CONST vfloat vmaxnum_vf_vf_vf(vfloat x, vfloat y) {
+  return vmax_vf_vf_vf(vsel_vf_vo_vf_vf(visnan_vo_vf(x), y, x), vsel_vf_vo_vf_vf(visnan_vo_vf(y), x, y));
+}
+
+static INLINE CONST vfloat vminnum_vf_vf_vf(vfloat x, vfloat y) {
+  return vmin_vf_vf_vf(vsel_vf_vo_vf_vf(visnan_vo_vf(x), y, x), vsel_vf_vo_vf_vf(visnan_vo_vf(y), x, y));
```

Comment:
Apologies for raising another question. This is the original implementation of the code that is currently in `sleefsimdsp.c`. For the x86 helper files, you have changed such implementation with `return vsel_vf_vo_vf_vf(visnan_vo_vf(y), x, vmin_vf_vf_vf(x, y));`. The new one seems to be better, would it be possible to replace this old one with the new one? If the answer is yes, I would suggest you to make the generic implementation in `df.h` as `return vsel_vf_vo_vf_vf(visnan_vo_vf(y), x, vmin_vf_vf_vf(x, y));` and remove the code (`#ifdef`s and function definitions) you have added in the x86 helper files. This reasoning about vmin holds also for the vmax function and the double precision versions in `dd.h`.

Reply:
I'm not sure I follow. This is the original  (i.e. currently in master) implementation in `sleefsimdsp.c`:
```
EXPORT CONST vfloat xfminf(vfloat x, vfloat y) {
  return vsel_vf_vo_vf_vf(visnan_vo_vf(y), x, vsel_vf_vo_vf_vf(vgt_vo_vf_vf(y, x), x, y));
}
```
The original version is not really generic -- it's tuned to the semantics of the x86 `minps/maxps/vminps/vmaxps` instructions; that's why the new x86-specific versions of `vminnum/vmaxnum` are slightly-optimized versions of the original, with `vsel(vgt())` changed to `vmin()`.

The proposed in this PR generic versions of `vminnum/vmaxnum` found in `df.h/dd.h` are actually generic, i.e. agnostic of the peculiarities of the underlying min/max vector ops. The proposed implementation of generic `vminnum/vmaxnum` in `df.h/dd.h` and specialized versions in the respective ISA helpers is the most logical implementation of this functionality.

Reply:
The original implementation is optimized in terms of the number of operations. When min and max instructions are available, I think NAN handling is not required. So, it just needs one instruction, like the following.
```
EXPORT CONST vdouble xfmax(vdouble x, vdouble y) {
   return vmax_vd_vd_vd(x, y);
}
```

Reply:
@shibatch, aren't `xfmax()/xfmin()` supposed to implement IEEE754-2008 minNum/maxNum semantics? The way the original is implemented, it tries to adhere to minNum/maxNum assuming x86 unerlying semantics (which are no natively minNum/maxNum-compliant).

Here's quick example of how the original would behave:
```
x: num, y: nan
original_xfmin(x, y) -> x

x: nan, y: num
original_xfmin(x, y) -> y
```
In contrast to a x86 `vmin()`:
```
x: num, y: nan
x86_vmin(x, y) -> nan

x: nan, y: num
x86_vmin(x, y) -> y
```

Reply:
NAN handling should be implemented in those instructions.

http://www.felixcloutier.com/x86/MAXPD.html

Reply:
While x86 `minps/maxps/vminps/vmaxps` have some handling of NaNs, that handling is not IEEE754-2008/C99-compliant. To quote the semantics from that page your referred to:

>If only one value is a NaN (SNaN or QNaN) for this instruction, the second operand (source operand), either a NaN or a valid floating-point value, is written to the result.

In contrast to C99 (http://en.cppreference.com/w/c/numeric/math/fmin):
>If one of the two arguments is NaN, the value of the other argument is returned. Only if both arguments are NaN, NaN is returned

Reply:
Okay, I now understand. You are correct.

Reply:
But the instructions in ARM is IEEE754-compliant.

http://infocenter.arm.com/help/index.jsp?topic=/com.arm.doc.dui0802b/CIHFCJCF.html

If we seriously trying to optimize this, it is better to make it choose the implementation according to IEEE754-compliance of the instructions.
I think this is a good opportunity to do this.

Reply:
@shibatch , are you saying that we should get completely rid of the old implementation and use the ones that @blu added in the x86 helper header files as generic implementation to be used in `[dd|df].h`?

This makes sense to me too.

Reply:
@fpetrogalli-arm, looking again at the new generic version, I don't think it would schedule better on any of the current popular uarchs than the original `xfmin()/xfmax()`. So I think the original `xfmin()/xfmax()` should become the generic version; the one issue with the new x86 version becoming the generic is that it won't work on ISAs with fmin/fmax semantics returning NaN when either of the args is NaN.

Reply:
In a situation like this, I usually write the exported functions (xfmin()/xfmax() in this case) in a generic way. Here, being generic means that the function returns the correct value regardless of the architecture. If I write an optimized version that works only for specific vector extensions, it is chosen with #if macro.

Basically, the exported functions (those functions beginning with x) should not be called by other implementation of functions. I also implement functions for internal usage, which do not handle non-numbers.

Reply:
@shibatch @fpetrogalli-arm Ok, give me a couple of hours (currently preoccupied) to show you what I mean.